packages/suite/src/middlewares/suite/metadataMiddleware.ts

```diff
@@ -1,35 +1,56 @@
 import { MiddlewareAPI } from 'redux';
 
-import { accountsActions, deviceActions } from '@suite-common/wallet-core';
-
 import * as metadataLabelingActions from 'src/actions/suite/metadataLabelingActions';
 import { AppState, Action, Dispatch } from 'src/types/suite';
-import { METADATA_LABELING, ROUTER } from 'src/actions/suite/constants';
+import { ROUTER, SUITE } from 'src/actions/suite/constants';
+import { selectMetadata } from 'src/reducers/suite/metadataReducer';
+import { discoveryActions, selectDiscoveryByDeviceState } from '@suite-common/wallet-core';
+import { ExperimentalFeature } from 'src/constants/suite/experimental';
 
 const metadata =
     (api: MiddlewareAPI<Dispatch, AppState>) =>
     (next: Dispatch) =>
     (action: Action): Action => {
-        if (accountsActions.createAccount.match(action)) {
-            action.payload = api.dispatch(
-                metadataLabelingActions.setAccountMetadataKey(action.payload),
-            );
-        }
+        const prevState = api.getState().metadata.entities || [];
 
         // pass action
         next(action);
 
-        if (deviceActions.receiveAuthConfirm.match(action)) {
-            if (
-                action.payload.success &&
-                api.getState().metadata.enabled &&
-                !action.payload.device.metadata[METADATA_LABELING.ENCRYPTION_VERSION]
-            ) {
-                api.dispatch(metadataLabelingActions.init(false));
+        switch (action.type) {
+            // detect changes in state in labelable entities.
+            // if labelable entitities differ from previous state after discovery completed init metadata
+
+            case discoveryActions.completeDiscovery.type:
+            case discoveryActions.updateDiscovery.type: {
+                // note: we presume here that discovery runs always only for currently selcted device
+                const { deviceState } = action.payload;
+                const discovery = selectDiscoveryByDeviceState(api.getState(), deviceState);
+                if (!discovery) {
+                    break;
+                }
+                const metadata = selectMetadata(api.getState());
+                if (!metadata.enabled || metadata.initiating) {
+                    break;
+                }
+
+                if (discovery.status === 4 && !discovery.authConfirm) {
+                    const nextState = api.dispatch(
+                        metadataLabelingActions.getLabelableEntitiesDescriptors(),
+                    );
+                    if (nextState.length && prevState.join('') !== nextState.join('')) {
```

Comment:
I absolutely don't understand this condition. Todo: figure out/explain/refactord/or at least add some comments